custom/abdm/auth.py

```diff
@@ -0,0 +1,19 @@
+from rest_framework.authentication import TokenAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+from custom.abdm.models import ABDMUser
+
+
+class ABDMUserAuthentication(TokenAuthentication):
+
+    def authenticate(self, request):
+        access_token = request.META.get('HTTP_AUTHORIZATION', "")
+        token = access_token.split(" ")[-1]
+        if not token:
+            return None
+        try:
+            user = ABDMUser.objects.get(access_token=token)
+        except ABDMUser.DoesNotExist:
+            raise AuthenticationFailed('Unauthorized')
+        if not user.is_token_valid:
+            raise AuthenticationFailed('Invalid Token')
```

Comment:
We are not passing the username with the requests, so Unauthorized will come when token is wrong. Second exception is to tell that token is correct but not valid anymore, such as expiry case. We do not have expiry mechanism currently so it will always return `True`. For case of invalid token, app has support added to notify user to sync again from the server.

Reply:
I see.

In what case the `ABDMUser` would not exist? Isn't that always present?

Still not sure about letting the user know that token as specifically expired. Could we generally tell the user that "The token was incorrect/expired. Please sync to get a fresh token"?

Reply:
1. `ABDMUser` will be created reactively first time when a sync request is sent by mobile worker. If no sync done then request will not have token and this error will be raised.
2. We can take this as an enhancement as I believe ABDM app is currently making use of these messages currently to display message to user in more descriptive way.

Reply:
Won't the mobile app be missing a token, if it tries to use the ABDM integration but has not synced it.
Not sure why HQ needs to highlight that.

Reply:
>  highlight that
Which part are we talking about here?

Reply:
Highlighting the user that HQ doesn't have a corresponding ABDMUser for them.

The other workaround would be for the mobile app to tell the user to sync if they are missing a token itself. Isn't that possible? I am guessing that this change would be needed in the CommCare app which wont be easy to get done?

Reply:
> tell the user to sync if they are missing a token itself
I think this is done in the ABDM app